### PR TITLE
Integrate fixes for Timestamp[MICROS] and infinite loop hang when reading parquet

### LINKED ISSUE #1458: Hang due to infinite loop when reading some parquet files with RLE encoding and bit packing
**Describe the bug**
Some parquet writers produce RLE encoded dictionaries with bit packing even if value count is not divisible by 8. 
This violates the assumption decoder makes about the number of values present in the bit-packed run, causing infinite loop and hang.

**To Reproduce**
Attempt to read parquet file where number of values is not power of 2 but RLE encoding with bit packing was used.

**Expected behavior**
File should be read correctly.

**Additional context**

### LINKED ISSUE #1459: Handle Parquet Files With Inconsistent Timestamp Units
**Describe the bug**
If parquet is written with timestamps with time unit other than `ns` reading such file would produce incorrect dates, whereas pandas is reading the dates correctly 

**To Reproduce**
Generate parquet file as follows:
`
import pandas as pd
import numpy as np

np.random.seed(0)
# create an array of 5 dates starting at '2015-02-24', one per minute
rng = pd.date_range('2020-01-01', periods=5, freq='H')
df = pd.DataFrame({ 'Date': rng, 'Val': np.random.randn(len(rng)) })
df.to_parquet('data/myfile.parquet', coerce_timestamps='ms', allow_truncated_timestamps=True)
`

**Expected behavior**
Data is not corrupted and dates are read back correctly.

**Additional context**
_

### PR DESCRIPTION
**Pull request by anliakho2**
# Which issue does this PR close?

Closes #1459 , closes #1458 .

# Rationale for this change
 
Fixing few issues experiences reading parquet files 

# What changes are included in this PR?

Add handling for the Timestamp(TimeUnit::MICROS) and Timestamp(TimeUnit::MILLIS) logical types.
Also handle RLE bit packed runs that don't have proper full packing. 

# Are there any user-facing changes?

no, except that more files would be read correctly.

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
Thank you @anliakho2 

Can you please add some regression tests / update existing ones to cover the missed cases? I would like to ensure we don't accidentally break these features in the future

File: parquet/src/arrow/array_reader.rs [hunk 1]
```diff
@@ -428,6 +430,24 @@ where
...
+                            let a = arrow::compute::cast(&array, &ArrowType::Timestamp(Microsecond, tz.clone()))?;
+                            arrow::compute::cast(&a, &ArrowType::Timestamp(Nanosecond, tz.clone()))?
+                        }
+                        TimeUnit::MILLIS(_) => {
+                            let a = arrow::compute::cast(&array, &ArrowType::Timestamp(Millisecond, tz.clone()))?;
+                            arrow::compute::cast(&a, &ArrowType::Timestamp(Nanosecond, tz.clone()))?
+                        }
+                        _ => arrow::compute::cast(&array, &target_type.clone())?
+                    }
+                }
```
Comment (viirya) [inline parquet/src/arrow/array_reader.rs:446]: The logic looks correct. But we need some tests to verify them.

**Comment by anliakho2** [comment]
@viirya @alamb Will do tests shortly, need some time to finish up  tasks at work :)

**Comment by alamb** [comment]
Hi @anliakho2  -- just checking in on this one. I plan to cut a new arrow release next week (Thursday or Friday) and it would be great to include this issue as well.

**Comment by tustvold** [comment]
I would be happy to add some tests / fix the merge conflicts if you would like a hand getting this over the line. As @alamb says, it would be awesome to include this in the next release :smile:

**Comment by anliakho2** [comment]
@alamb Sorry was travelling and just came back... I think I could get back to it in time. @tustvold I would also appreciate your help on this, so feel free

**Comment by tustvold** [comment]
Following investigation in #1459 I think we want to handle this differently, I will close this, cherry-pick the fix for #1458 into a separate PR and get another PR up to alter how we infer the schema of a parquet file to be compatible with what I think is a bug in pyarrow (https://issues.apache.org/jira/browse/ARROW-16184)

**Pull request closed without merging**